incubator/hnc/internal/validators/hierarchy.go

```diff
@@ -210,9 +212,55 @@ func (v *Hierarchy) checkParent(ns, curParent, newParent *forest.Namespace) admi
 		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
 	}
 
+	// Prevent overwriting source objects in the descendants after the hierarchy change.
+	if yes, co := v.hasConflictObject(newParent, ns); yes {
+		coStr := strings.Join(co, "\n  * ")
+		msg := fmt.Sprintf("\nCannot update hierarchy because it would overwrite the following object(s):\n  * %s\nTo fix this, please rename or remove the conflicting objects first.", coStr)
+		return deny(metav1.StatusReasonConflict, msg)
+	}
+
 	return allow("")
 }
 
+// hasConflictObject returns true and a list of objects if there's a conflict.
+func (v *Hierarchy) hasConflictObject(pns, cns *forest.Namespace) (bool, []string) {
+	// Get all types with 'Propagate' mode.
+	ts := []forest.TypeSyncer{}
+	for _, t := range v.Forest.GetTypeSyncers() {
+		if t.GetMode() == api.Propagate {
+			ts = append(ts, t)
+		}
+	}
```

Comment:
This code might be more clear if you split it into two functions: one to iterate over the GVKs, and another to construct the conflicting strings just for that GVK:

```
func (v *Hierarchy) getConflictingObjects(pns, cns *forest.Namespace) []string {
  var conflicts []string
  for _, t := range v.Forest.GetTypeSyncers() {
    if t.GetMode() == api.Propagate {
      conflicts = append(conflicts, v.getConflictingObjectsOfType(t.GetGVK(), pns, cns))
    }
  }
  return conflicts
}
```

Then your lower-level functions only need to deal with one type, which should make them much simpler.